test/suites/container_devices.sh

```diff
@@ -0,0 +1,105 @@
+  ensure_import_testimage
+  ensure_has_localhost_remote "${LXD_ADDR}"
+
+  veth_host_name="veth$$"
+  ct_name=c1
+  ipRand=$(shuf -i 0-9 -n 1)
+
+  #Test pre-launch profile config is applied at launch.
+  lxc profile device set default eth0 ipv4.routes "192.0.2.1${ipRand}/32"
+  lxc profile device set default eth0 ipv6.routes "2001:db8::1${ipRand}/128"
+  lxc profile device set default eth0 limits.ingress 1Mbit
+  lxc profile device set default eth0 limits.egress 2Mbit
+  lxc profile device set default eth0 host_name "${veth_host_name}"
+  lxc launch testimage "${ct_name}"
+  if ! ip -4 r | grep "192.0.2.1${ipRand}" ; then
+    echo "ipv4.routes invalid"
+    exit 1
+  fi
+  if ! ip -6 r | grep "2001:db8::1${ipRand}" ; then
+    echo "ipv4.routes invalid"
+    exit 1
+  fi
+  if ! tc class show dev "${veth_host_name}" | grep "1Mbit" ; then
+    echo "limits.ingress invalid"
+    exit 1
+  fi
+  if ! tc filter show dev "${veth_host_name}" egress | grep "2Mbit" ; then
+    echo "limits.egress invalid"
+    exit 1
+  fi
+
+  #Test hot plugging a container nic with different settings to profile with the same name.
+  lxc config device add "${ct_name}" eth0 nic \
+    nictype=bridged \
+    parent=lxdbr0 \
+    ipv4.routes="192.0.2.2${ipRand}/32" \
+    ipv6.routes="2001:db8::2${ipRand}/128" \
+    limits.ingress=3Mbit \
+    limits.egress=4Mbit \
+    host_name="${veth_host_name}"
+
+  if ! ip -4 r | grep "192.0.2.2${ipRand}" ; then
+    echo "ipv4.routes invalid"
+    exit 1
+  fi
+  if ! ip -6 r | grep "2001:db8::2${ipRand}" ; then
+    echo "ipv4.routes invalid"
+    exit 1
+  fi
+  if ! tc class show dev "${veth_host_name}" | grep "3Mbit" ; then
+    echo "limits.ingress invalid"
+    exit 1
+  fi
+  if ! tc filter show dev "${veth_host_name}" egress | grep "4Mbit" ; then
+    echo "limits.egress invalid"
+    exit 1
+  fi
+
+  #Test removing hot plugged device and check profile nic is restored.
+  lxc config device remove "${ct_name}" eth0
+  if ! ip -4 r | grep "192.0.2.1${ipRand}" ; then
+    echo "ipv4.routes invalid"
+    exit 1
+  fi
+  if ! ip -6 r | grep "2001:db8::1${ipRand}" ; then
+    echo "ipv4.routes invalid"
+    exit 1
+  fi
+  if ! tc class show dev "${veth_host_name}" | grep "1Mbit" ; then
+    echo "limits.ingress invalid"
+    exit 1
+  fi
+  if ! tc filter show dev "${veth_host_name}" egress | grep "2Mbit" ; then
+    echo "limits.egress invalid"
+    exit 1
+  fi
+
+  #Test hot plugging a container nic then updating it.
+  lxc config device add "${ct_name}" eth0 nic \
+    nictype=bridged \
+    parent=lxdbr0 \
+    host_name="${veth_host_name}"
+  lxc config device set "${ct_name}" eth0 ipv4.routes "192.0.2.2${ipRand}/32"
+  lxc config device set "${ct_name}" eth0 ipv6.routes "2001:db8::2${ipRand}/128"
+  lxc config device set "${ct_name}" eth0 limits.ingress 3Mbit
+  lxc config device set "${ct_name}" eth0 limits.egress 4Mbit
+
+  if ! ip -4 r | grep "192.0.2.2${ipRand}" ; then
+    echo "ipv4.routes invalid"
+    exit 1
```

Comment:
Do not call `exit` from tests, just call `false`.

Calling exit would bypass all our cleanup code.

Reply:
(Fix all the other ones too :))